src/indexer/index_writer.rs

```diff
@@ -1048,4 +1057,21 @@ mod tests {
         assert_eq!(num_docs_containing("b"), 0);
         fail::cfg("RAMDirectory::atomic_write", "off").unwrap();
     }
+
+    #[test]
```

Comment:
missing unit test.


#  test_add_then_clear.
- what happens if within a commit, 
   1) I add a doc
   2 ) I clear the index
   3) I commit
I would expect the resulting index to contain no docs. (I don't think this is the case right now).

# test_clear_then_add

- what happens if within a commit
   1) I clear the index
   2 ) I add a doc
   3) I commit

I would expect the document to be here.

# test_clear_and_rollback

1) add docs
2) commit
3) I clear the index
4) I rolback

I would expect the index to contain original docs.

Reply:
`test_add_then_clear` might be tricky to implement. 

The most efficient way to do it might be to `rollback`.
For tracing purpose, I'd prefer if you pulled all of the code except logging into another private method of IndexWriter,
Something like `rollback_impl()` and call `rollback_impl()` at the beginning of `.clear()` and in `.rollback()`.

So `fn rollback(&mut self)` would look something like 

```rust
fn rollback(&mut self) {
     info!("Rolling back to opstamp {}", self.committed_opstamp);
     self.rollback_impl();
     Ok(())
}
```

One thing that sucks with this implementation, is that we reset the opstamp counter too. We might need a bit of refactoring to do things properly.